docker_cli/_cli.py

```diff
@@ -0,0 +1,327 @@
+import argparse
+import sys
+import textwrap
+from typing import Sequence
+
+from ._utils import universal_tag_prefix
+from .commands import dropin, make_lockfile, remove
+from .setup_commands import (setup_all, setup_cuda_dev, setup_cuda_runtime,
+                             setup_env_add, setup_env_dev, setup_env_runtime,
+                             setup_init)
+
+
+def setup_parser() -> argparse.ArgumentParser:
```

Comment:
What would you say are the benefits of doing this in an object-oriented way like you've described above, over the procedural method of splitting these into a series of functions that each do part of the work?

Reply:
My goal in the above was just to impose some more structure on the code to make it a bit more organized and (IMO) easier to read. Reading over the code the other day, I found that it was hard to see at a glance which parameters belonged to which subcommands and it seemed like there was too much going on the `setup_parser()` function to fully comprehend.

The object-oriented approach seemed like an intuitive strategy to apply here, since each subcommand needs to provide at least two pieces of information: (1) how to parse its command-line arguments, and (2) how to run the command with the parsed arguments. The most obvious way to do that to me seemed to be to make those two methods in a class.

The object-oriented way also makes it easy to add new subcommands in the future by providing a clear interface that subcommands must conform to in order to plug in to the framework.

That said, this was just intended to be one example of a possible approach. If you have a better strategy in mind that's more procedural, I'd be very happy to take a look. My main goal is just that we break up `setup_parser()` into smaller, more manageable pieces.

Reply:
I'm thinking let's stick with the procedural implementation and move the incomplete OOP version to a backup branch somewhere to potentially revisit down the line.

TBH the procedural version still seems a bit less modular than I'd ideally want, but I think further changes in this section right now would potentially cause a lot of churn in subsequent PRs that you've got lined up and are pretty low-priority for me. So I'm imagining we might pick this discussion back up later on, but the procedural approach is fine with me for now.